### PR TITLE
[load test: warn] [MC-1229] feat: add metrics to curated-recommendations

### PR DESCRIPTION
**Pull request by mmiermans**
## References

JIRA: [MC-1229](https://mozilla-hub.atlassian.net/browse/MC-1229)
Graphana: [Merino Curated Recommendations dashboard](https://earthangel-b40313e5.influxcloud.net/d/nqIyrq_Sk/merino-py-curated-recommendations?orgId=1&var-environment=stagepy&var-canary=All&from=1721076381565&to=1721077470717)

## Description
Create a Graphana dashboard for the curated-recommendations endpoint, to be used in addition to the [existing Merino dashboard](https://earthangel-b40313e5.influxcloud.net/d/rQAfYKIVk/merino-py-application-and-infrastructure?orgId=1&refresh=1m). We want to start simple with metrics that are essential for monitoring after the roll-out to release, and for analyzing a load test.

## Implementation decision
I think the following are essential:
- Endpoint latency (p50, p95, p99)
- Endpoint status codes
- Curated-corpus-api latency (p50, p95, p99)
- Curated-corpus-api status codes

image

Source: [Merino Curated Recommendations dashboard](https://earthangel-b40313e5.influxcloud.net/d/nqIyrq_Sk/merino-py-curated-recommendations?orgId=1&var-environment=stagepy&var-canary=All&from=1721076381565&to=1721077470717)

## Deployment

After deployment, check and create panels in Graphana for metrics created in this PR. It's expected that it shows "No data" until this is merged.

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [x] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [x] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [x] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [ ] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [x] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

[MC-1229]: https://mozilla-hub.atlassian.net/browse/MC-1229?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: merino/curated_recommendations/corpus_backends/corpus_api_backend.py [hunk 1]
```diff
@@ -187,10 +197,17 @@ async def _fetch_from_backend(self, surface_id: ScheduledSurfaceId) -> list[Corp
...
+            "curated_recommendations.corpus_api_backend.request.timing"
+        ):
+            res = await self.http_client.post(
+                self.graph_config.endpoint,
+                json=body,
+                headers=self.graph_config.headers,
+            )
+
+        self.metrics_client.increment(
+            f"curated_recommendations.corpus_api_backend.request.status_codes.{res.status_code}"
```
Comment (mmiermans) [inline merino/curated_recommendations/corpus_backends/corpus_api_backend.py:210]: @ncloudioj Do these metric names look good to you? I tried to follow Merino conventions.

File: tests/integration/api/v1/curated_recommendations/test_curated_recommendations.py [hunk 2]
```diff
@@ -383,3 +390,36 @@ async def test_cache_returned_on_subsequent_calls(
...
             assert all("NEW" in item["title"] for item in new_data["data"])
+
+
+@pytest.mark.asyncio
+async def test_curated_recommendations_metrics(
+    mocker: MockerFixture,
+) -> None:
+    """Test that metrics are recorded for the 'suggest' endpoint
+    (status codes: 200 & 400).
+    """
```
Comment (Trinaa) [inline tests/integration/api/v1/curated_recommendations/test_curated_recommendations.py:401]:
```suggestion
    """Test that metrics are recorded for the 'curated_recommendations_metrics' endpoint (status codes: 200).
    """
```

**Comment by efixler** [comment]
Hey @mmiermans this looks really good! I think you made a good decision here including the corpus latency. Not a requirement (could be handled in a follow-up) but corpus errors would probably be a good thing too!

**Review by jpetto — COMMENTED** [review]
i like how easy it is to add to the grafana dashboard - i'd love to get something like this in our monorepo.

agree with @efixler that it would be good to get corpus errors in here as well if possible (potentially as a follow up), and with @Trinaa's comment on the test.

looking good though!

File: merino/curated_recommendations/corpus_backends/corpus_api_backend.py (hunk 1 above)
Comment (ncloudioj) [reply merino/curated_recommendations/corpus_backends/corpus_api_backend.py:208]: Looks good to me. We normally prefer shorter IDs (so long as no ambiguity they will have) to ease the work in Grafana. For example, `recommend.corpus_api.request.status_code.{res.status_code}`

File: docs/data.md [hunk 3]
```diff
@@ -69,6 +69,13 @@ log inspection interfaces.
 - `ERROR merino.featureflags` - There was an error while attempting to assign a
   feature flag for a suggest API request.
 
+### Curated Recommendations
```
Comment (ncloudioj) [inline docs/data.md:72]: Thanks for documenting the new log and metrics!

**Review by ncloudioj — APPROVED** [review]

**Review by katerinachinnappan — APPROVED** [review]
love how straightforward it is to add a grafana dashboard! agreed with Jonathan, something for us to think about adding to the monorepo. ty for also adding 1265 in this PR.

**Comment by mmiermans** [comment]
> Hey @mmiermans this looks really good! I think you made a good decision here including the corpus latency. Not a requirement (could be handled in a follow-up) but corpus errors would probably be a good thing too!

Thanks! Let me add an item to [MC-1270](https://mozilla-hub.atlassian.net/browse/MC-1270) to track GraphQL errors. Note that this PR already tracks 4xx and 5xx status codes.

File: tests/integration/api/v1/curated_recommendations/test_curated_recommendations.py [hunk 4]
```diff
@@ -115,271 +122,336 @@ async def test_curated_recommendations():
...
-            assert corpus_http_client.post.call_count == 2
-            new_data = new_response.json()
-            assert new_data["recommendedAt"] > initial_data["recommendedAt"]
-            assert all("NEW" in item["title"] for item in new_data["data"])
+            with pytest.raises(HTTPStatusError):
+                await fetch_en_us(ac)
+
+            # TODO: Remove reliance on internal details of aiodogstatsd
+            metric_keys_cache_miss: list[str] = [call.args[0] for call in report.call_args_list]
+            assert metric_keys_cache_miss == [
```
Comment (jpetto) [inline tests/integration/api/v1/curated_recommendations/test_curated_recommendations.py:449]: why are these `metric_keys_cache_` vars named differently? e.g. `_hit` and `_miss`?  do those different suffixes mean something?

Comment (mmiermans) [reply tests/integration/api/v1/curated_recommendations/test_curated_recommendations.py:449]: Good point. That's an artifact from these tests being previously combined in one. I'll remove the suffix.

**Pull request merged**